File: Muse/Muse WatchKit Extension/InterfaceController.m
//
//  InterfaceController.m
//  Muse WatchKit Extension
//
//  Created by Jieyi Hu on 6/28/15.
//  Copyright © 2015 SenseWatch. All rights reserved.
//

#import "InterfaceController.h"


@interface InterfaceController()
@property WCSession *session;
@property CMMotionManager* cmMotionManager;
@property NSTimeInterval lastGestureTimestamp;
@end


@implementation InterfaceController

- (void)awakeWithContext:(id)context {
    [super awakeWithContext:context];

    // Configure interface objects here.
}

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
    [super willActivate];
    [self setUpProperties];
}

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
//    [self.cmMotionManager stopAccelerometerUpdates];
    [super didDeactivate];
}

- (IBAction)lastSongButtonClick {
    NSLog(@"lastSongButton On Watch is clicked.");
    [self sendGesture:@"LEFT"];
}

- (IBAction)playSongButtonClick {
    NSLog(@"playSongButton On Watch is clicked.");
    [self sendGesture:@"PUSH"];
}

- (IBAction)nextSongButtonClick {
    NSLog(@"nextSongButton On Watch is clicked.");
    [self sendGesture:@"RIGHT"];
}

- (void)setUpProperties{
    self.lastGestureTimestamp = 0;
    [self setUpWCSession];
    [self setUpMontionSensor];
}

- (void)setUpWCSession{
    _session = [WCSession defaultSession];
    _session.delegate = self;
    [_session activateSession];
    NSLog(@"WCSession On Watch is activated.");
}

- (void)setUpMontionSensor{
    _cmMotionManager = [[CMMotionManager alloc] init];
    NSLog(@"CMMontionManager on Watch is initialized.");
    if([self.cmMotionManager isAccelerometerAvailable]){
        NSLog(@"Accelerometer on Watch is available.");
        _cmMotionManager.accelerometerUpdateInterval = 0.1;
        [self.cmMotionManager startAccelerometerUpdatesToQueue:[[NSOperationQueue alloc] init] withHandler:^(CMAccelerometerData* data,NSError* error){
            dispatch_async(dispatch_get_main_queue(), ^{
                //  Handler Code
                [_dataLabel setText:[NSString stringWithFormat:@"x:%.02f   y:%.02f   z:%.02f",data.acceleration.x,data.acceleration.y,data.acceleration.z]];
                [self handleAccelerometerData:data];
            });
        }];

    } else{
        NSLog(@"Accelerometer on Watch is not available.");
    }
}

- (void)handleAccelerometerData:(CMAccelerometerData*)data{
    float x = data.acceleration.x;
    float y = data.acceleration.y;
    float z = data.acceleration.z;
    NSString* gestureName = @"null";

    if(sqrt(x*x+y*y+z*z) > 1.25){
        //  has gesture made
        if(z > y){
            //  gesture UP
            gestureName = @"UP";
        } else if(fabs(z) > fabs(x)){
            //  gesture DOWN/PUSH
            gestureName = @"PUSH";
        } else{
            //  gesture LEFT/RIGHT
            gestureName = @"RIGHT";
        }
    } else{
        //  do nothing
    }
    
    if([gestureName isEqualToString:@"null"]){
        //no gesture made
    } else{
        //  there is gesture made
        if(self.lastGestureTimestamp == 0){
            //  there is no last gesture
            [self sendGesture:gestureName];
            _lastGestureTimestamp = data.timestamp;
        } else{
            //  there is a last gesture
            if(data.timestamp - self.lastGestureTimestamp >= 1){
                //  last gesture was made at least one sec ago
                [self sendGesture:gestureName];
                _lastGestureTimestamp = data.timestamp;
            } else{
                //  do nothing
            }
        }
    }
}


- (void)sendGesture:(NSString*)gesture{
    if([_session isReachable]){
        [_session sendMessage:@{@"type":@"Gesture",
                                @"content":gesture} replyHandler:nil errorHandler:nil];
        NSLog(@"Gesture on Watch is sent");
    } else{
        NSLog(@"WCSession on Watch is not reachable");
    }
}

- (void)session:(nonnull WCSession *)session didReceiveMessage:(nonnull NSDictionary<NSString *,id> *)message{
    NSString* type = [message objectForKey:@"type"];
    if([type isEqualToString:@"UIUpdateInfo"]){
        NSLog(@"Watch receives UI update info");
        [self handleUIUpdateInfo:[message objectForKey:@"content"]];
    }
}

- (void)handleUIUpdateInfo:(NSDictionary*)updateInfo{
    [[WKInterfaceDevice currentDevice] playHaptic:WKHapticTypeClick];
    [self.songTitleLabel setText:[updateInfo objectForKey:@"nowPlayingSongTitle"]];
    [self.playSongButton setBackgroundImage:[UIImage imageNamed:[updateInfo objectForKey:@"playButtonImageName"]]];
}

@end

File: AirFlip/AirFlip WatchKit Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  BSSDemo WatchKit Extension
//
//  Created by Jieyi Hu on 7/24/15.
//  Copyright © 2015 SenseWatch. All rights reserved.
//
import UIKit
import WatchKit
import Foundation
import CoreMotion


class InterfaceController: WKInterfaceController {
    
    var hardwareCtrl = HKWatchController()
    let mgr = CMMotionManager()
    
    // MARK: Notification related
    var lastNotificationDate = NSDate(timeIntervalSince1970: 0)
    let notificationPushInterval : NSTimeInterval = 1.0
    
    
    @IBOutlet var motionSensorSwitch: WKInterfaceSwitch!
    
    @IBAction func pushButtonClick() {
        hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Push.notification)
    }
    @IBAction func leftButtonClick() {
        hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Left.notification)
    }
    @IBAction func rightButtonClick() {
        hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Right.notification)
    }
    @IBAction func upButtonClick() {
        hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Up.notification)
    }
    @IBAction func downButtonClick() {
        hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Down.notification)
    }
    @IBAction func motionSensorSwitchClick(value: Bool) {
        if value {
            mgr.accelerometerUpdateInterval = 0.01
            mgr.startAccelerometerUpdatesToQueue(NSOperationQueue(), withHandler: { (data, error) -> Void in
                if let data = data {
                    let x = data.acceleration.x
                    let y = data.acceleration.y
                    let z = data.acceleration.z
                    if sqrt(x*x+y*y+z*z) > 1.35 {
                        if self.lastNotificationDate.timeIntervalSinceNow <= -self.notificationPushInterval {
                            if z > y {
                                self.hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Up.notification)
                            } else {
                                self.hardwareCtrl.comSession.sendNotification(GKPredefinedGesture.Down.notification)
                            }
                            self.lastNotificationDate = NSDate()
                        }
                    }
                }
            })
        } else {
            mgr.stopAccelerometerUpdates()
        }
    }
    
    override func awakeWithContext(context: AnyObject?) {
        super.awakeWithContext(context)
        
        // Configure interface objects here.
    }
    
    override func willActivate() {
        // This method is called when watch view controller is about to be visible to user
        super.willActivate()
    }
}
